scripts/testing/e2e-runner

```diff
@@ -0,0 +1,396 @@
+            return 0
+            ;;
+    esac
+
+    if [ ! -e "$CLONED_REPO" ]; then
+        must-mkdir "$CLONED_REPO" "local repo"
+        must-cd "$CLONED_REPO"
+        if ! git clone "$REMOTE_REPO" .; then
+            fatal "failed to clone $REMOTE_REPO to $CLONED_REPO"
+        fi
+    else
+        must-cd "$CLONED_REPO"
+        if ! git remote -v | grep origin | grep -q "$REMOTE_REPO"; then
+            fatal "existing clone $CLONED_REPO has other origin than $REMOTE_REPO"
+        fi
+        if ! git fetch origin; then
+            fatal "failed to fetch $REMOTE_REPO"
+        fi
+    fi
+
+    must-mkdir "$worktree_dir" "worktree root"
+
+    WORKTREE="$(realpath $WORKTREE)"
+    info "setting up git worktree $WORKTREE..."
+    if ! git worktree add "$WORKTREE" "origin/$TEST_BRANCH"; then
+        fatal "failed to add git worktree $WORKTREE"
+    fi
+
+    if [ -x "$WORKTREE/scripts/testing/e2e-runner" ]; then
+        exec "$WORKTREE/scripts/testing/e2e-runner" \
+             --skip-worktree \
+             ${PROXIES_SRC:+--source-proxies }${PROXIES_SRC:-} \
+             --remote "$REMOTE_REPO" \
+             --branch "$TEST_BRANCH" \
+             --local "$CLONED_REPO" \
+             --results "$RESULT_ROOT" \
+             --name "$RESULT_NAME" \
+             "$E2E_TESTS"
+    else
+        must-cd "$WORKTREE"
+    fi
+}
+
+prepare_result_dir() {
+    local dir="$RESULT_DIR"
+    local log="$dir/${SELF##*/}.log"
+
+    info "creating result dir $dir..."
+    must-mkdir "$dir" "result"
+
+    if ! touch "$log"; then
+        fatal "failed to create log file $log"
+    fi
+    LOG="$log"
+
+    info "Redirecting output to log file $LOG..."
+    if ! exec &>"$LOG"; then
+        fatal "log redirection failed"
+    fi
+}
+
+run_e2e_tests() {
+    local tests
+
+    info "running e2e tests ${E2E_TESTS:-with the default test set}..."
+    must-cd "$WORKTREE"
+    if [ -n "$E2E_TESTS" ]; then
+        make E2E_TESTS="$E2E_TESTS" e2e-tests
+    else
+        make e2e-tests
+    fi
+}
+
+cleanup_test_vms() {
+    local src
+
+    info "cleaning up test VMs..."
+    must-cd "$WORKTREE"
+    for src in test/e2e/n[0-9]*c[0-9]*-c*; do
+        (cd "$src" && vagrant destroy --force)
+    done
+}
+
+collect_git_info() {
+    local describe="$RESULT_DIR/$GIT_DESC"
+    local sha1="$RESULT_DIR/$GIT_SHA1"
+
+    info "collecting info about worktree..."
+    must-cd "$WORKTREE"
+    git describe > "$describe"
+    git log -1 --pretty=%H > "$sha1"
+}
+
+collect_results() {
+    local src dst
+
+    info "collecting test results..."
+    must-cd "$WORKTREE"
+    for src in test/e2e/n[0-9]*c[0-9]*-c*; do
+        dst=$RESULT_DIR/${src#test/e2e/}
+        must-mkdir "$dst" "result"
+        must-mv "$src/$TEST_SUITE_DIR" "$dst"
+    done
+}
+
+cleanup_worktree() {
+    info "cleaning up worktree $WORKTREE..."
+    must-cd "$WORKTREE"
+    if ! git clean -fdx; then
+        fatal "failed to git clean worktree $WORKTREE"
+    fi
```

Comment:
Why clean it as we're removing it anyway? Just do `git worktree remove -f` below?

```suggestion
```

Reply:
I'm fine with updating it like that, although I think it is not exactly the same (worktree remove --force will also nuke a locked directory while git clean -fdx is at least not documented to do the same) but I don't think that matters here.

Reply:
What locking are you referring to? Locking worktrees or smth else?